dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,446 @@
+      Lisha Li, Kevin Jamieson, Giulia DeSalvo, Afshin Rostamizadeh, Ameet Talwalkar
+      https://arxiv.org/abs/1603.06560
+
+    """
+    def __init__(self, model, params, max_iter=81, eta=3,
+                 n_jobs=-1, **kwargs):
+        """
+        Parameters
+        ----------
+        params :
+            Search space of various parameters. This mirrors the sklearn API.
+        model :
+            The base model to try.
+        max_iter : int
+            The maximum number of iterations to train a model on, as determined
+            by ``model.partial_fit``.
+        X, y : np.ndarray, np.ndarray
+            Train data, features ``X`` and labels ``y``.
+        eta : optional, int
+            How aggressive to be while search. The default is 3, and theory
+            suggests that ``eta=e=2.718...``. Changing this is not recommended.
+        n_jobs : int, optional
+            Hyperband depends on another function, and they are completely
+            indepdenent of one another and can be run embarassingly parallel.
+            This function is submitted to the scheduler if ``n_jobs == -1``
+            (which is the default behavior). Setting ``n_jobs == 0`` can help
+            preserve memory on your machine.
+        """
+        self.params = params
+        self.model = model
+        self.R = max_iter
+        self.eta = eta
+        self.best_val_score = -np.inf
+        self.n_iter = 0
+
+        if n_jobs not in {-1, 0}:
+            raise ValueError('n_jobs must be either -1 or 0')
+        self.n_jobs = n_jobs
+
+        est = model if not isinstance(model, ParallelPostFit) else model.estimator
+        if not hasattr(est, 'partial_fit'):
+            raise ValueError('Hyperband relies on partial_fit. Without it '
+                             'it would be no different than RandomizedSearchCV')
+        if not hasattr(est, 'warm_start'):
+            warnings.warn('model has no attribute warm_start. Hyperband will assume it '
+                          'is reusing previous calls to `partial_fit` during each call '
+                          'to `partial_fit`')
+        else:
+            if not est.warm_start:
+                raise ValueError('Hyperband relies on model(..., warm_start=True).')
+
+        self.history = []
+
+        super_default = dict(scoring=None, iid=True, refit=True, cv=None,
+                             error_score='raise', return_train_score=_RETURN_TRAIN_SCORE_DEFAULT,
+                             scheduler=None, cache_cv=True)
+        for k, v in super_default.items():
+            if k in kwargs and kwargs[k] != super_default[k]:
+                warnings.warn('Hyperband ignores the keyword argument {k}.')
+        super(Hyperband, self).__init__(model, **kwargs)
+
+    def fit(self, X, y, dry_run=False, **fit_kwargs):
+        """
+        This function requires an active dask.distribtued client as it uses
+        the dask.distributed's concurrent.futures API.
```

Comment:
Also, just to motivate use of the local scheduler, it would be motivational to be able to look at the graph produced by a Hyperband stage and also quite useful to run it in a single thread for profiling and benchmarking purposes.

Reply:
I'll investigate visualizing the graph in computation; I think it'll be a good exercise.

I'm inclined to believe the graph does change during execution.   Several calls to `_successive_halving` are made  in parallel from `Hyperband.fit`, and each of those makes an adaptive choice on which models to train further.

To verify my understanding: if each of the `_successive_halving` calls were in serial (say, in a vanilla list-comphrension), then the task graph would not change. It would be created many times from the scheduler, but it would be static each time because it is only concerned with fitting different models, not making an adaptive choice.  Does this sound correct?

Reply:
> I'm inclined to believe the graph does change during execution.

You make several independent calls to `dask.compute`.  My understanding is that at any time when dask is active the graph does not change.  Because of this I don't think we need to use anything fancy, like the concurrent.futures or joblib interfaces.  I believe that you can do everything you want to do here with the local scheduler.

Reply:
Got it – the tree of `compute` calls independent and not intertwined. I've refactored to use dask.delayed, and it's working.